synapse/tests/utils.py

```diff
@@ -481,6 +481,46 @@ def thisHostMustNot(self, **props):  # pragma: no cover
             if s_thishost.get(k) == v:
                 raise unittest.SkipTest('skip thishost: %s==%r' % (k, v))
 
+    def get_form_test_data(self, core, formname):
+        t = core.model.type(formname)
+        f = core.model.form(formname)
+
+        valu_repr = t.info.get('ex')
+        valu_norm = t.norm(valu_repr)[0]
+        expected_ndef = (formname, valu_norm)
+
+        input_props = {}
+        expected_props = {}
+        for pname in f.props:
+
+            # Skip univs
+            if pname.startswith('.'):
+                continue
+
+            prop = f.prop(pname)
+            ex_repr = prop.info.get('ex', prop.type.info.get('ex', prop.type.opts.get('ex')))
```

Comment:
would want to clean this up and write out how the example resolution order works